### PR TITLE
Switching deprecated buffer constructor to use Buffer.from

### PR DESCRIPTION
**Pull request by micahriggan**
* new Buffer() is deprecated
  * https://nodejs.org/api/buffer.html#buffer_new_buffer_array
* new Buffer([]) should be Buffer.from([])
* new Buffer(9) should be Buffer.alloc(9)
* BufferWriter should keep track of length as you write
* Transaction.hash getter should cache the last hash, until you change it.

Before Changes 
```
Benchmarking Script
---------------------------------------
isPublicKeyHashIn x 42,112,577 ops/sec ±2.14% (421 runs sampled)
isPublicKeyOut x 30,459,210 ops/sec ±2.80% (430 runs sampled)
toAddress x 65,897 ops/sec ±3.15% (323 runs sampled)
getAddressInfo x 86,997 ops/sec ±2.68% (325 runs sampled)
Done
----------------------------------------------------------------------
Finished
```

After Changes 
```
Benchmarking Script
---------------------------------------
isPublicKeyHashIn x 65,906,640 ops/sec ±0.98% (516 runs sampled)
isPublicKeyOut x 51,775,114 ops/sec ±0.92% (510 runs sampled)
toAddress x 122,002 ops/sec ±1.03% (437 runs sampled)
getAddressInfo x 143,790 ops/sec ±0.93% (528 runs sampled)
Done
----------------------------------------------------------------------
Finished
```

### CODE REVIEW

File: lib/transaction/transaction.js [hunk 1]
```diff
@@ -838,6 +842,7 @@ Transaction.prototype._getInputAmount = function() {
 };
 
 Transaction.prototype._updateChangeOutput = function() {
+  this._hashCache = false;
```
Comment (matiu) [inline lib/transaction/transaction.js]: Shouldn't we set hashCache  to false in other function, like: .sign?

Comment (nitsujlangston) [reply lib/transaction/transaction.js]: let's take the hash cache part out of this PR - need to discuss plan on that.

Comment (micahriggan) [reply lib/transaction/transaction.js]: Good idea!

File: lib/encoding/bufferwriter.js [hunk 2]
```diff
@@ -22,58 +24,60 @@ BufferWriter.prototype.toBuffer = function() {
...
 
 BufferWriter.prototype.concat = function() {
-  return Buffer.concat(this.bufs);
+  return Buffer.concat(this.bufs, this.bufLen);
 };
 
 BufferWriter.prototype.write = function(buf) {
   assert(bufferUtil.isBuffer(buf));
   this.bufs.push(buf);
+  this.bufLen += buf.length;
```
Comment (micahriggan) [inline lib/encoding/bufferwriter.js]: Do we still want to do this one?

**Pull request merged**